File: command_static/src/panels/cutting_interval.js
/* Realised cutting interval against target.
 *
 * The realised complement of the rotation panel. The rotation panel says which
 * blocks are due against a synthetic target; this one measures, from the
 * client's own cutting days, how long the round actually ran on every block
 * and how it drifted from January to May. The intervals are real; the target
 * and the gang are synthetic, and the screen says which is which.
 */
import { esc, fmt } from '../lib/fmt.js';
import { blockList } from './_shared.js';

// Inline-block: `.kpi span` is display:block, which would turn the badge into a bar.
const REAL = '<span class="prov real" style="display:inline-block;margin-left:4px" title="From the client\'s EPMS harvest export">real</span>';
const SYN = '<span class="prov synthetic" style="display:inline-block;margin-left:4px" title="Generated: gis/data/synthetic/ec_rotation.csv">synthetic</span>';

const n1 = v => v === null || v === undefined ? '—' : fmt(v, Number.isInteger(v) ? 0 : 1);
const days = v => v === null || v === undefined ? '—' : `${n1(v)} d`;
const x = v => v === null || v === undefined ? '—' : `${Number(v).toFixed(2)}×`;
const pc = v => v === null || v === undefined ? '—' : `${Math.round(v)}%`;
const shortDate = s => {
  if (!s) return '—';
  const d = new Date(`${s}T00:00:00`);
  return d.toLocaleDateString('en-GB', { day: 'numeric', month: 'short' });
};

/* Monthly median and P90 round as an SVG line chart, with the synthetic
   target's median as a dashed reference and the Lebaran stop marked. */
function roundChart(series, targetMedian, closure) {
  if (!series || series.length < 2) return '';
  const W = 640, H = 210, L = 38, R = 18, T = 18, B = 30;
  const n = series.length;
  const top = Math.max(...series.map(s => Math.max(s.p90 || 0, s.median || 0)), targetMedian || 0);
  const yMax = Math.ceil((top + 3) / 5) * 5;
  const xAt = i => L + (W - L - R) * (n === 1 ? 0.5 : i / (n - 1));
  const yAt = v => T + (H - T - B) * (1 - v / yMax);
  const ticks = [];
  for (let t = 0; t <= yMax; t += 5) ticks.push(t);
  const pts = key => series.map((s, i) => `${xAt(i).toFixed(1)},${yAt(s[key] || 0).toFixed(1)}`);
  const band = [...pts('p90'), ...pts('median').reverse()].join(' ');
  const txt = 'fill:var(--muted);font-size:10.5px;font-family:var(--font)';

  // The stop fell in the last days of March; mark it just before April's point.
  let lebaran = '';
  if (closure && closure.from) {
    const mi = series.findIndex(s => s.month === closure.from.slice(0, 7));
    if (mi >= 0 && mi + 1 < n) {
      const xm = xAt(mi) + (xAt(mi + 1) - xAt(mi)) * 0.87;
      lebaran = `<line x1="${xm.toFixed(1)}" x2="${xm.toFixed(1)}" y1="${T}" y2="${H - B}"
          style="stroke:var(--text);stroke-width:1;stroke-dasharray:2 3;opacity:.6"/>
        <text x="${(xm - 4).toFixed(1)}" y="${T + 10}" text-anchor="end" style="${txt};fill:var(--text)">Lebaran stop, ${closure.days} d</text>`;
    }
  }
  const target = targetMedian ? `
    <line x1="${L}" x2="${W - R}" y1="${yAt(targetMedian).toFixed(1)}" y2="${yAt(targetMedian).toFixed(1)}"
        style="stroke:var(--gold);stroke-width:1.5;stroke-dasharray:6 4"/>
    <text x="${W - R}" y="${(yAt(targetMedian) - 5).toFixed(1)}" text-anchor="end"
        style="${txt};fill:var(--gold)">synthetic target, median ${n1(targetMedian)} d</text>` : '';

  return `<svg class="ci-chart" viewBox="0 0 ${W} ${H}" role="img"
      style="display:block;width:100%;height:auto;max-width:760px;margin:4px 0 2px"
      aria-label="Median and P90 cutting round by month">
    ${ticks.map(t => `<line x1="${L}" x2="${W - R}" y1="${yAt(t).toFixed(1)}" y2="${yAt(t).toFixed(1)}"
        style="stroke:var(--border-s);stroke-width:1"/>
      <text x="${L - 6}" y="${(yAt(t) + 3).toFixed(1)}" text-anchor="end" style="${txt}">${t}</text>`).join('')}
    ${series.map((s, i) => `<text x="${xAt(i).toFixed(1)}" y="${H - 9}"
        text-anchor="${i === 0 ? 'start' : i === n - 1 ? 'end' : 'middle'}" style="${txt}">${esc(s.label)}${s.partial ? ' (to 23rd)' : ''}</text>`).join('')}
    <polygon points="${band}" style="fill:rgba(63,176,196,.16);stroke:none"/>
    <polyline points="${pts('p90').join(' ')}" style="fill:none;stroke:var(--accent);stroke-width:1;stroke-dasharray:3 3;opacity:.8"/>
    <polyline points="${pts('median').join(' ')}" style="fill:none;stroke:var(--accent);stroke-width:2;stroke-linejoin:round"/>
    ${series.map((s, i) => `<circle cx="${xAt(i).toFixed(1)}" cy="${yAt(s.median || 0).toFixed(1)}" r="3" style="fill:var(--accent)"/>
      <text x="${xAt(i).toFixed(1)}" y="${(yAt(s.median || 0) - 8).toFixed(1)}" text-anchor="middle"
        style="${txt};fill:var(--text);font-weight:600">${n1(s.median)}</text>`).join('')}
    ${target}
    ${lebaran}
  </svg>
  <div class="st-legend"><span><i class="lg-mid"></i>median round</span>
    <span><i class="lg-band"></i>median to P90</span>
    <span><i class="lg-rop"></i>synthetic target</span></div>`;
}

export async function panelCuttingInterval() {
  const r = await fetch('/gis/cutting-interval?estate=EC&top=12');
  if (!r.ok) return `<div class="empty">Cutting interval endpoint answered ${r.status}.</div>`;
  const d = await r.json();
  if (!d.available) return `<div class="empty">${esc(d.reason || 'No cutting-interval data.')}</div>`;

  const t = d.totals, tg = d.target || {}, s = d.series || [];
  const anchor = shortDate(d.window && d.window.anchor);
  const monthKeys = s.map(m => m.month);

  return `<div class="sheet-note" style="margin:0 0 12px;padding:0;border:none;color:var(--text);font-size:12.5px">
    ${esc(d.summary)}</div>

  <div class="kpis">
    <div class="kpi"><b>${days(t.early_round)}</b><span>round, Jan–Feb ${REAL}</span></div>
    <div class="kpi"><b>${days(t.late_round)}</b><span>round, Apr–May ${REAL}</span></div>
    <div class="kpi"><b>${x(t.stretch)}</b><span>stretch ${REAL}</span></div>
    <div class="kpi"><b>${t.gap_over_14_blocks}</b><span>blocks &gt;14 d uncut at ${esc(anchor)} ${REAL}</span></div>
    <div class="kpi"><b>${pc(t.beyond_14_pct)}</b><span>rounds over 14 d ${REAL}</span></div>
    <div class="kpi"><b>${pc(t.beyond_target_pct)}</b><span>rounds beyond target ${SYN}</span></div>
  </div>

  ${roundChart(s, tg.median, d.closure)}

  <table class="tbl">
    <tr><th>Month</th><th class="num">Rounds</th><th class="num">Median</th><th class="num">P90</th>
        <th class="num">Max</th><th class="num">Over 14 d</th><th class="num">Beyond target</th></tr>
    ${s.map(m => `<tr class="${(m.median || 0) > 14 ? 'hi' : ''}">
      <td>${esc(m.label)}${m.partial ? ' <span style="color:var(--muted)">to 23rd</span>' : ''}</td>
      <td class="num">${m.intervals}</td>
      <td class="num"><b>${days(m.median)}</b></td>
      <td class="num">${days(m.p90)}</td>
      <td class="num">${days(m.max)}</td>
      <td class="num">${pc(m.beyond_14_pct)}</td>
      <td class="num" style="color:var(--gold)">${pc(m.beyond_target_pct)}</td>
    </tr>`).join('')}
  </table>
  <div class="blk-cap">Each round is filed under the month the return cut fell in. ${REAL} throughout;
    the target column is the only synthetic figure.</div>

  <div class="sub-t">By division</div>
  <table class="tbl">
    <tr><th>Div</th><th class="num">Blocks</th>
        ${monthKeys.map((k, i) => `<th class="num">${esc(s[i].label.slice(0, 3))}</th>`).join('')}
        <th class="num">Stretch</th><th class="num">&gt;14 d uncut</th></tr>
    ${(d.by_division || []).map(v => `<tr class="${(v.stretch || 0) >= 2 ? 'hi' : ''}">
      <td>${esc(v.division_code)}</td><td class="num">${v.blocks}</td>
      ${monthKeys.map(k => `<td class="num">${n1(v.months[k])}</td>`).join('')}
      <td class="num"><b>${x(v.stretch)}</b></td>
      <td class="num">${v.gap_over_14}</td>
    </tr>`).join('')}
  </table>

  <div class="sub-t">Blocks whose round stretched most</div>
  ${blockList(d.most_stretched, {
    cols: [
      { key: 'division_code', label: 'Div' },
      { key: 'early_round', label: 'Jan–Feb', num: true, fmt: days },
      { key: 'late_round', label: 'Apr–May', num: true, fmt: days },
      { key: 'stretch_vs_own', label: 'Stretch', num: true, fmt: x },
      { key: 'target', label: 'Target', num: true, fmt: days },
      { key: 'beyond_target_pct', label: 'Beyond target', num: true, fmt: pc },
    ],
    caption: 'Stretch is the block\'s April–May round over its own January–February round: real against real. '
      + 'Target and beyond-target are against the synthetic round.',
  })}

  <div class="sub-t">Longest without a cut at ${esc(anchor)}</div>
  ${blockList(d.longest_gap, {
    cols: [
      { key: 'division_code', label: 'Div' },
      { key: 'gap_at_anchor', label: 'Days uncut', num: true, fmt: days },
      { key: 'last_cut', label: 'Last cut', fmt: shortDate },
      { key: 'median_round', label: 'Median round', num: true, fmt: days },
      { key: 'target', label: 'Target', num: true, fmt: days },
      { key: 'synthetic_days_since', label: 'Rotation panel says', num: true, fmt: days },
    ],
    caption: `Days uncut is the export's last day less the block's last real cutting day. `
      + `The rotation panel's figure follows the synthetic ledger and disagrees on ${tg.last_cut_disagrees_blocks} of ${t.blocks} blocks.`,
  })}

  ${(d.by_gang || []).length ? `<div class="sub-t">By gang ${SYN}</div>
  <table class="tbl">
    <tr><th>Gang</th><th class="num">Blocks</th><th class="num">Jan–Feb</th><th class="num">Apr–May</th><th class="num">Stretch</th></tr>
    ${d.by_gang.slice(0, 6).map(g => `<tr>
      <td>${esc(g.gang_code)}</td><td class="num">${g.blocks}</td>
      <td class="num">${days(g.early_round)}</td><td class="num">${days(g.late_round)}</td>
      <td class="num"><b>${x(g.stretch)}</b></td></tr>`).join('')}
  </table>
  <div class="blk-cap">The rounds are real; the gang on each block is invented, so this attribution is illustrative only.</div>` : ''}

  <div class="sheet-note">
    ${d.closure ? `<b>Lebaran.</b> ${esc(d.closure.note)}<br><br>` : ''}
    <b>The target.</b> ${esc(tg.note || '')}<br><br>
    <b>Caveat.</b> ${esc(d.caveat)}<br><br>
    <b>What the client learns on their own data.</b> ${(d.learns && d.learns.on_own_data || []).map(esc).join(' ')}<br>
    <b>What the synthetic layer adds.</b> ${(d.learns && d.learns.synthetic_adds || []).map(esc).join(' ')}<br><br>
    <b>Provenance.</b> ${esc(d.provenance)}. ${esc(d.note)}
  </div>`;
}

File: gis/cutting_interval.py
"""Realised cutting interval against target.

Answers "is the round actually being kept, block by block?" from the client's
own harvest export. The EPMS OPH snapshot (gis/data/estates/EC/oph.csv, from
gis/build_estate_data.py) carries one or more rows per block per harvest
date, so the distinct dates per block are the block's real cutting days, and
the interval between successive cuts is the block's realised round. That
figure is REAL.

Two things about the raw days matter and are handled here rather than left to
the reader:

  * A block is not cut in one day. A gang works a block over two or three
    consecutive days and comes back a week or more later, so the raw
    day-to-day gaps are bimodal: 1-2 days (the same visit continuing) and
    7-15 days (the round). The round is therefore measured between VISITS - a
    visit being a run of cutting days with at most one rest day between them -
    from the first day of one visit to the first day of the next.

  * The synthetic target in ec_rotation.csv was built on a different basis:
    calendar days per cutting day over January-February, clamped to 5-14. A
    block worked two consecutive days every ten gets a target of 5 while its
    round is 10. The target is still reported and compared as asked, but the
    like-for-like baseline is the block's own January-February round, which is
    real, and the split is explicit in the payload.

Loaded once and cached; the endpoint answers from memory after the first call.
"""
from __future__ import annotations

import csv
import logging
import statistics
from collections import defaultdict
from datetime import date, timedelta
from pathlib import Path
from threading import Lock

from gis import layers

log = logging.getLogger("estate-command.cutting_interval")

_ESTATES_DIR = Path(__file__).resolve().parent / "data" / "estates"
_CACHE: dict = {}
_LOCK = Lock()

# A cutting day within `REST_DAYS` of the previous one belongs to the same
# visit. One allows the Sunday break (Saturday -> Monday) without splitting a
# visit; the gap distribution has its valley at 4-5 days, so anything wider is
# the gang coming back for the next round.
REST_DAYS = 1
# The early-year window the synthetic target claims to reproduce, and the
# late window the trend is measured against.
EARLY_TO = date(2025, 2, 28)
LATE_FROM = date(2025, 4, 1)
# The round ceiling the synthetic target band tops out at. Not a client
# standard; used only to count rounds that ran long on the real data.
LONG_ROUND = 14

_MONTH_NAMES = {1: "January", 2: "February", 3: "March", 4: "April", 5: "May",
                6: "June", 7: "July", 8: "August", 9: "September",
                10: "October", 11: "November", 12: "December"}


# ── loading ────────────────────────────────────────────────────────────────

def _cutting_days(estate: str) -> dict[str, list[date]] | None:
    """Real cutting days per block, from the client's OPH export.

    Cut at the synthetic window's end: the target it is compared against, and
    the early/late windows above, belong to that window. The snapshot itself
    now runs well past it.
    """
    code = estate.upper()
    path = _ESTATES_DIR / code / "oph.csv"
    if not path.exists():
        return None
    through = layers.EXPORT_END
    days: dict[str, set] = defaultdict(set)
    with path.open(encoding="utf-8-sig", newline="") as fh:
        rd = csv.reader(fh)
        head = next(rd)
        # EPMS spells this column 'divison_code' in the export. Kept as-is.
        i_date = head.index("harvest_date")
        i_div = head.index("divison_code")
        i_blk = head.index("block_code")
        for row in rd:
            if row[i_date] > through:
                continue
            days[layers._key(row[i_div], row[i_blk])].add(row[i_date])
    return {k: sorted(date.fromisoformat(d) for d in s) for k, s in days.items()}


def _visits(days: list[date]) -> list[tuple[date, date]]:
    """Runs of cutting days with at most REST_DAYS idle days between them."""
    out: list[list[date]] = [[days[0], days[0]]]
    for d in days[1:]:
        if (d - out[-1][1]).days <= REST_DAYS + 1:
            out[-1][1] = d
        else:
            out.append([d, d])
    return [(a, b) for a, b in out]


def _median(vals):
    return round(statistics.median(vals), 1) if vals else None


def _p90(vals):
    if not vals:
        return None
    s = sorted(vals)
    return s[max(0, -(-9 * len(s) // 10) - 1)]   # nearest-rank P90


def _month_label(m: str) -> str:
    return _MONTH_NAMES[int(m[5:7])]


def _compute(estate: str) -> dict:
    days_by_block = _cutting_days(estate)
    if not days_by_block:
        return {"available": False,
                "reason": f"No harvest export for {estate.upper()}: expected "
                          f"gis/data/estates/{estate.upper()}/oph.csv - run "
                          f"gis/build_estate_data.py."}

    rows = layers.block_rows(estate, synthetic_world=True) or []
    meta = {layers._key(r["division_code"], r["block_code"]): r for r in rows}
    rot = layers._state()["rotation"]
    anchor = date.fromisoformat(layers.EXPORT_END)

    lo = min(d[0] for d in days_by_block.values())
    hi = max(d[-1] for d in days_by_block.values())

    # Estate-wide closure: calendar days on which no block anywhere was cut.
    cut_any = {d for ds in days_by_block.values() for d in ds}
    closed = []
    d = lo
    while d <= hi:
        if d not in cut_any:
            closed.append(d)
        d += timedelta(days=1)
    # The longest closure is the one that stretches rounds. In 2025 that is
    # Lebaran: Idul Fitri fell on 31 March and the estate stopped 27-31 March.
    runs: list[list[date]] = []
    for c in closed:
        if runs and (c - runs[-1][-1]).days == 1:
            runs[-1].append(c)
        else:
            runs.append([c])
    longest = max(runs, key=len) if runs else []

    blocks = []
    all_iv: list[dict] = []
    for k, days in days_by_block.items():
        m = meta.get(k, {})
        r = rot.get(k) or {}
        vis = _visits(days)
        ivs = []
        for (a0, a1), (b0, b1) in zip(vis, vis[1:]):
            ivs.append({"days": (b0 - a0).days, "from": a0, "to": b0,
                        "straddles": bool(longest) and a1 < longest[0] and b0 > longest[-1]})
        n = [iv["days"] for iv in ivs]
        early = [iv["days"] for iv in ivs if iv["to"] <= EARLY_TO]
        late = [iv["days"] for iv in ivs if iv["to"] >= LATE_FROM]
        target = r.get("target")
        early_med = _median(early) if len(early) >= 2 else None
        late_med = _median(late) if len(late) >= 1 else None
        med = _median(n)
        last_cut = days[-1]
        blocks.append({
            "block_id": m.get("block_id"),
            "block_label": m.get("block_label") or k.replace("|", "-"),
            "division_code": m.get("division_code") or k.split("|")[0],
            "block_code": k.split("|")[1],
            "planted_ha": m.get("planted_ha"),
            "gang_code": r.get("gang_code"),
            "cutting_days": len(days),
            "visits": len(vis),
            "median_visit_days": _median([(b - a).days + 1 for a, b in vis]),
            "intervals": len(n),
            "median_round": med,
            "p90_round": _p90(n),
            "max_round": max(n) if n else None,
            "early_round": early_med,
            "late_round": late_med,
            "stretch_vs_own": (round(late_med / early_med, 2)
                               if early_med and late_med else None),
            "target": target,
            "stretch_vs_target": (round(med / target, 2) if med and target else None),
            "beyond_target_pct": (round(100.0 * sum(1 for x in n if x > target) / len(n), 1)
                                  if n and target else None),
            "beyond_14_pct": (round(100.0 * sum(1 for x in n if x > LONG_ROUND) / len(n), 1)
                              if n else None),
            "last_cut": last_cut.isoformat(),
            "gap_at_anchor": (anchor - last_cut).days,
            "synthetic_days_since": r.get("days_since"),
            "synthetic_last_cut": r.get("last_harvest_date"),
        })
        for iv in ivs:
            all_iv.append({**iv, "k": k, "div": blocks[-1]["division_code"],
                           "target": target})

    blocks.sort(key=lambda b: (b["division_code"], int(b["block_code"])))

    # ── by month: the round closed in that month ─────────────────────────
    months = sorted({iv["to"].strftime("%Y-%m") for iv in all_iv})
    series = []
    for mo in months:
        ivm = [iv for iv in all_iv if iv["to"].strftime("%Y-%m") == mo]
        vals = [iv["days"] for iv in ivm]
        # The buildplan's basis, kept so its figure can be checked: calendar
        # days in the month over the block's cutting days in it.
        y, mm = int(mo[:4]), int(mo[5:7])
        m_start = date(y, mm, 1)
        m_end = min(hi, date(y + (mm == 12), mm % 12 + 1, 1) - timedelta(days=1))
        dpc = []
        for ds in days_by_block.values():
            c = sum(1 for x in ds if m_start <= x <= m_end)
            if c:
                dpc.append(((m_end - m_start).days + 1) / c)
        straddle = [iv["days"] for iv in ivm if iv["straddles"]]
        series.append({
            "month": mo,
            "label": _month_label(mo),
            "intervals": len(vals),
            "blocks": len({iv["k"] for iv in ivm}),
            "median": _median(vals),
            "p90": _p90(vals),
            "max": max(vals) if vals else None,
            "beyond_target_pct": round(100.0 * sum(1 for iv in ivm if iv["target"] and iv["days"] > iv["target"]) / len(ivm), 1) if ivm else None,
            "beyond_14_pct": round(100.0 * sum(1 for x in vals if x > LONG_ROUND) / len(vals), 1) if vals else None,
            "days_per_cutting_day": _median(dpc),
            "straddling_closure": len(straddle),
            "median_straddling": _median(straddle),
            "median_not_straddling": _median([iv["days"] for iv in ivm if not iv["straddles"]]),
            "partial": m_end < date(y + (mm == 12), mm % 12 + 1, 1) - timedelta(days=1),
        })

    divs = sorted({b["division_code"] for b in blocks}, key=int)
    by_division = []
    for dv in divs:
        ivd = [iv for iv in all_iv if iv["div"] == dv]
        e = [iv["days"] for iv in ivd if iv["to"] <= EARLY_TO]
        l = [iv["days"] for iv in ivd if iv["to"] >= LATE_FROM]
        em, lm = _median(e), _median(l)
        by_division.append({
            "division_code": dv,
            "blocks": sum(1 for b in blocks if b["division_code"] == dv),
            "months": {mo: _median([iv["days"] for iv in ivd
                                    if iv["to"].strftime("%Y-%m") == mo]) for mo in months},
            "early_round": em,
            "late_round": lm,
            "stretch": round(lm / em, 2) if em and lm else None,
            "beyond_target_pct": round(100.0 * sum(1 for iv in ivd if iv["target"] and iv["days"] > iv["target"]) / len(ivd), 1) if ivd else None,
            "gap_over_14": sum(1 for b in blocks if b["division_code"] == dv
                               and b["gap_at_anchor"] > LONG_ROUND),
        })

    # Gang attribution rides on the invented roster; kept apart and labelled.
    by_gang: dict = defaultdict(lambda: {"blocks": 0, "early": [], "late": []})
    for b in blocks:
        if not b["gang_code"]:
            continue
        g = by_gang[b["gang_code"]]
        g["blocks"] += 1
        if b["early_round"]:
            g["early"].append(b["early_round"])
        if b["late_round"]:
            g["late"].append(b["late_round"])
    gangs = []
    for code, g in sorted(by_gang.items()):
        em, lm = _median(g["early"]), _median(g["late"])
        gangs.append({"gang_code": code, "blocks": g["blocks"], "early_round": em,
                      "late_round": lm, "stretch": round(lm / em, 2) if em and lm else None})
    gangs.sort(key=lambda g: -(g["stretch"] or 0))

    # ── totals ───────────────────────────────────────────────────────────
    all_days = [iv["days"] for iv in all_iv]
    early_all = [iv["days"] for iv in all_iv if iv["to"] <= EARLY_TO]
    late_all = [iv["days"] for iv in all_iv if iv["to"] >= LATE_FROM]
    with_target = [iv for iv in all_iv if iv["target"]]
    targets = [b["target"] for b in blocks if b["target"]]
    early_rounds = [b["early_round"] for b in blocks if b["early_round"]]
    em, lm = _median(early_all), _median(late_all)
    stretch = round(lm / em, 2) if em and lm else None
    beyond_target = (round(100.0 * sum(1 for iv in with_target if iv["days"] > iv["target"])
                           / len(with_target), 1) if with_target else None)
    beyond_14 = round(100.0 * sum(1 for x in all_days if x > LONG_ROUND) / len(all_days), 1)
    gap_over_14 = [b for b in blocks if b["gap_at_anchor"] > LONG_ROUND]
    target_med = _median(targets)
    early_block_med = _median(early_rounds)
    at_floor = sum(1 for t in targets if t == 5)
    last_mismatch = sum(1 for b in blocks if b["synthetic_last_cut"]
                        and b["synthetic_last_cut"] != b["last_cut"])
    latest = series[-1] if series else None
    first = series[0] if series else None

    most_stretched = sorted([b for b in blocks if b["stretch_vs_own"]],
                            key=lambda b: (-b["stretch_vs_own"], -(b["late_round"] or 0)))
    longest_gap = sorted(blocks, key=lambda b: (-b["gap_at_anchor"], -(b["median_round"] or 0)))

    summary = (
        f"The round has stretched. Blocks were cut about every {em:g} days in "
        f"January-February and every {lm:g} days in April-May, "
        f"{stretch:g} times as long; {len(gap_over_14)} of {len(blocks)} blocks "
        f"had gone more than {LONG_ROUND} days without a cut at "
        f"{anchor.strftime('%d %B %Y').lstrip('0')}."
        if em and lm and stretch and stretch >= 1.15 else
        f"The round is holding. Blocks were cut about every {em:g} days in "
        f"January-February and every {lm:g} days in April-May; "
        f"{len(gap_over_14)} of {len(blocks)} blocks had gone more than "
        f"{LONG_ROUND} days without a cut at {anchor.strftime('%d %B %Y').lstrip('0')}."
    )

    closure = None
    if longest:
        straddlers = [iv for iv in all_iv if iv["straddles"]]
        closure = {
            "from": longest[0].isoformat(), "to": longest[-1].isoformat(),
            "days": len(longest),
            "intervals_straddling": len(straddlers),
            "median_straddling": _median([iv["days"] for iv in straddlers]),
            "note": (f"No block anywhere was cut {longest[0].strftime('%d %b').lstrip('0')}-"
                     f"{longest[-1].strftime('%d %b').lstrip('0')} {longest[-1].year}: the "
                     f"Lebaran break (Idul Fitri fell on 31 March 2025). "
                     f"{len(straddlers)} rounds straddle it and are up to "
                     f"{len(longest)} days longer for that reason alone; the stretch "
                     f"persists through April and May after the break, so the "
                     f"holiday explains the spike, not the trend."),
        }

    return {
        "available": True,
        "estate": estate.upper(),
        "summary": summary,
        "window": {"from": lo.isoformat(), "to": hi.isoformat(), "anchor": anchor.isoformat(),
                   "rest_days_within_visit": REST_DAYS,
                   "early": f"{lo.isoformat()}..{EARLY_TO.isoformat()}",
                   "late": f"{LATE_FROM.isoformat()}..{hi.isoformat()}"},
        "totals": {
            "blocks": len(blocks),
            "blocks_with_target": len(targets),
            "cutting_days": sum(b["cutting_days"] for b in blocks),
            "visits": sum(b["visits"] for b in blocks),
            "intervals": len(all_days),
            "median_round": _median(all_days),
            "p90_round": _p90(all_days),
            "max_round": max(all_days) if all_days else None,
            "median_visit_days": _median([b["median_visit_days"] for b in blocks
                                          if b["median_visit_days"]]),
            "early_round": em,
            "late_round": lm,
            "stretch": stretch,
            "first_month_median": first["median"] if first else None,
            "latest_month_median": latest["median"] if latest else None,
            "latest_month": latest["label"] if latest else None,
            "beyond_target_pct": beyond_target,
            "beyond_14_pct": beyond_14,
            "gap_over_14_blocks": len(gap_over_14),
            "gap_over_14_ha": round(sum(b["planted_ha"] or 0 for b in gap_over_14), 1),
            "median_gap_at_anchor": _median([b["gap_at_anchor"] for b in blocks]),
            "closed_days": [c.isoformat() for c in closed],
        },
        "target": {
            "source": "synthetic: gis/data/synthetic/ec_rotation.csv rotation_target_days",
            "basis": ("calendar days per cutting day over January-February, "
                      "clamped to 5-14"),
            "median": target_med,
            "at_floor_5": at_floor,
            "distribution": {str(t): targets.count(t) for t in sorted(set(targets))},
            "early_round_same_blocks": early_block_med,
            "runs_low_by_days": (round(early_block_med - target_med, 1)
                                 if early_block_med and target_med else None),
            "last_cut_disagrees_blocks": last_mismatch,
            "note": (f"The target counts calendar days per cutting day, not days "
                     f"between visits, so a block worked two days running every "
                     f"ten gets a target of 5 while its round is 10. Median target "
                     f"{target_med:g} against a median realised January-February "
                     f"round of {early_block_med:g} on the same blocks; "
                     f"{at_floor} of {len(targets)} targets sit on the 5-day floor. "
                     f"Read 'beyond target' with that in mind; the like-for-like "
                     f"test is each block against its own early-year round."),
        },
        "series": series,
        "by_division": by_division,
        "by_gang": gangs,
        "_most_stretched": most_stretched,
        "_longest_gap": longest_gap,
        "closure": closure,
        "buildplan_check": {
            "claimed": "median realised interval 5.2 days in January and 7.7 in May",
            "days_per_cutting_day": {s["label"]: s["days_per_cutting_day"] for s in series},
            "round_between_visits": {s["label"]: s["median"] for s in series},
            "verdict": ("The 5.2 and 7.7 reproduce exactly as calendar days per "
                        "cutting day, which is a work-rate figure. The round a "
                        "palm sees, visit to visit, is roughly twice that."),
        },
        "learns": {
            "on_own_data": [
                "The realised round per block, visit to visit, and how it drifted from January to May.",
                "Which blocks have gone longest without a cut at the export's last day.",
                f"The {len(longest) if longest else 0}-day Lebaran stop and how many rounds it stretched.",
                "Which divisions stretched most - the trend is real down to division level.",
            ],
            "synthetic_adds": [
                "A per-block target round to grade against. Invented, and built on a basis that runs low.",
                "A gang on each block, so the stretch can be attributed to a crew. Invented.",
                "A 'days since harvest' that disagrees with the client's own last cutting day on "
                f"{last_mismatch} of {len(blocks)} blocks; this panel uses the real day.",
            ],
        },
        "provenance": ("real: cutting days and every interval from the client's EPMS OPH "
                       "export; synthetic: the target round and the gang on each block"),
        "note": ("A visit is a run of cutting days with at most one rest day between "
                 "them; the round is measured from the first day of one visit to the "
                 "first day of the next. Each interval is filed under the month the "
                 "return cut fell in."),
        "caveat": ("May is cut short at the export's last day, so a round still open on "
                   f"{anchor.strftime('%d %B').lstrip('0')} is not in the May median; those "
                   "blocks show in the longest-gap list instead, which understates May. "
                   "The target's basis runs low against a visit-to-visit round, so "
                   "'beyond target' overstates; the block's own January-February round "
                   "is the fair yardstick. Sunday cutting runs at under half the weekday "
                   "rate, which the one-rest-day rule absorbs."),
    }


def position(estate: str = "EC", top: int = 12) -> dict:
    """Realised cutting interval per block against its target round."""
    code = estate.upper()
    with _LOCK:
        if code not in _CACHE:
            try:
                _CACHE[code] = _compute(code)
            except Exception as exc:  # a broken export must not take the panel down
                log.exception("[cutting_interval] %s failed", code)
                return {"available": False, "reason": f"cutting interval failed: {exc}"}
        d = _CACHE[code]
    if not d.get("available"):
        return d
    top = max(1, min(50, int(top or 12)))
    out = dict(d)
    # The cached lists are full; the endpoint trims to `top`.
    out["most_stretched"] = d["_most_stretched"][:top]
    out["longest_gap"] = d["_longest_gap"][:top]
    return {k: v for k, v in out.items() if not k.startswith("_")}


def reload() -> None:
    with _LOCK:
        _CACHE.clear()
